src/families/algorand/libcore-buildSubAccounts.js

```diff
@@ -0,0 +1,132 @@
+// @flow
+
+import type {
+  CryptoCurrency,
+  TokenAccount,
+  Account,
+  SyncConfig,
+} from "../../types";
+import type { CoreAccount } from "../../libcore/types";
+import { minimalOperationsBuilder } from "../../reconciliation";
+import { buildASAOperation } from "./buildASAOperation";
+import { BigNumber } from "bignumber.js";
+import { findTokenById, listTokensForCryptoCurrency } from "../../currencies";
+import { promiseAllBatched } from "../../promise";
+
+const OperationOrderKey = {
+  date: 0,
+};
+
+async function buildAlgorandTokenAccount({
+  parentAccountId,
+  token,
+  coreAccount,
+  existingTokenAccount,
+  balance,
+}) {
+  const id = parentAccountId + "+" + token.id;
+  const getAllOperations = async () => {
+    const query = await coreAccount.queryOperations();
+    await query.complete();
+    await query.addOrder(OperationOrderKey.date, false);
+    const coreOperations = await query.execute();
+
+    const operations = await minimalOperationsBuilder(
+      (existingTokenAccount && existingTokenAccount.operations) || [],
+      coreOperations,
+      (coreOperation) =>
+        buildASAOperation({
+          coreOperation,
+          accountId: id,
+        })
+    );
+
+    return operations;
+  };
+
+  const operations = await getAllOperations();
+
+  const tokenAccount: $Exact<TokenAccount> = {
+    type: "TokenAccount",
+    id,
+    parentId: parentAccountId,
+    starred: false,
+    token,
+    operationsCount: operations.length,
+    operations,
+    pendingOperations: [],
+    balance,
+    swapHistory: [],
+    creationDate:
+      operations.length > 0
+        ? operations[operations.length - 1].date
+        : new Date(),
+  };
+
+  return tokenAccount;
+}
+
+async function algorandBuildTokenAccounts({
+  currency,
+  coreAccount,
+  accountId,
+  existingAccount,
+  syncConfig,
+}: {
+  currency: CryptoCurrency,
+  coreAccount: CoreAccount,
+  accountId: string,
+  existingAccount: ?Account,
+  syncConfig: SyncConfig,
+}): Promise<?(TokenAccount[])> {
+  const { blacklistedTokenIds = [] } = syncConfig;
+  if (listTokensForCryptoCurrency(currency).length === 0) return undefined;
+  const tokenAccounts = [];
+  const algorandAccount = await coreAccount.asAlgorandAccount();
+  const accountASA = await algorandAccount.getAssetsBalances();
+
+  const existingAccountByTicker = {}; // used for fast lookup
+  const existingAccountTickers = []; // used to keep track of ordering
+  if (existingAccount && existingAccount.subAccounts) {
+    for (const existingSubAccount of existingAccount.subAccounts) {
+      if (existingSubAccount.type === "TokenAccount") {
+        const { ticker, id } = existingSubAccount.token;
+        if (!blacklistedTokenIds.includes(id)) {
+          existingAccountTickers.push(ticker);
+          existingAccountByTicker[ticker] = existingSubAccount;
+        }
+      }
+    }
+  }
+
+  // filter by token existence
+  promiseAllBatched(3, accountASA, async (asa) => {
```

Comment:
misses a `await` to not have race condition and/or finishing earlier than expected the `withLibcore` scope which can have bad consequences on mobile.

i highly suspect this will not work on mobile and was tested only on cli/desktop which is not actually asynchronous.

Reply:
We do not retest it on mobile will redo it